onnxruntime/core/framework/sparse_tensor.cc

```diff
@@ -3,31 +3,441 @@
 
 #include "core/framework/data_types.h"
 #include "core/framework/sparse_tensor.h"
+#include "core/framework/data_transfer_manager.h"
+#include "core/framework/utils.h"
+
+#include <safeint/SafeInt.hpp>
 
 using namespace onnxruntime::common;
 
 namespace onnxruntime {
 
-SparseTensor::SparseTensor(MLDataType elt_type,
-                           const TensorShape& shape,
-                           size_t nnz,
-                           void* values_data,
-                           void* indices_data,
-                           const OrtMemoryInfo& memory_info)
-    : values_(elt_type, TensorShape({static_cast<int64_t>(nnz)}), values_data, memory_info),
-      indices_(DataTypeImpl::GetType<int64_t>(),
-               TensorShape({static_cast<int64_t>(nnz), static_cast<int64_t>(shape.NumDimensions())}),
-               indices_data, memory_info, 0),
-      shape_(shape) {}
-
-SparseTensor::SparseTensor(MLDataType elt_type,
-                           const TensorShape& shape,
-                           size_t nnz,
+std::ostream& operator<<(std::ostream& os, SparseFormat flags) {
+  return os << std::hex << static_cast<uint32_t>(flags);
+}
+
+namespace {
+// Round up size to a multiple of align.
+// Example:
+//   Roundup(13, 5)   => 15
+//   Roundup(201, 16) => 208
+constexpr size_t kIndexAlignment = alignof(int64_t);
+int64_t Roundup(int64_t size) {
+  return ((SafeInt<int64_t>(size) + kIndexAlignment - 1) / kIndexAlignment) * kIndexAlignment;
+}
+
+/// <summary>
+/// Calculate required buffer size. We will place indices
+/// after data and make sure indices start at int64_t aligned place
+/// </summary>
+/// <returns></returns>
+int64_t CalculateRequiredBufferSize(int64_t data_size, int64_t indices_size) {
+  return SafeInt<int64_t>(Roundup(data_size)) + indices_size;
+}
+}  // namespace
+
+const void* SparseTensor::IndicesStart(int64_t values_bytes) const {
+  if (p_data_ != nullptr) {
+    return reinterpret_cast<const uint8_t*>(p_data_) + Roundup(values_bytes);
+  }
+  return nullptr;
+}
+
+void* SparseTensor::IndicesStart(int64_t values_bytes) {
+  if (p_data_ != nullptr) {
+    return reinterpret_cast<uint8_t*>(p_data_) + Roundup(values_bytes);
+  }
+  return nullptr;
+}
+
+int64_t SparseTensor::RequiredAllocationSize() const noexcept {
+  if (p_data_ != nullptr) {
+    // Can be zero for zero matrices
+    assert(buffer_size_ >= 0);
+    return buffer_size_;
+  } else {
+    auto data_size = values_.SizeInBytes();
+    int64_t index_size = 0;
+    for (const auto& t : format_data_) {
+      index_size += t.SizeInBytes();
+    }
+    return CalculateRequiredBufferSize(data_size, index_size);
```

Comment:
Does there need to be rounding for alignment of these individual items

Reply:
We only align the indices for after the values placement. Otherwise, we assume (just like Tensor) that allocators return a suitably aligned buffer. This function takes care of it.

Reply:
So if I understand it correctly, we have the values first, potential padding, followed by the first indicies Tensor. If it's CSR format there's no padding between the inner and the outer indices Tensor's. why do we not care about padding in that case?

Reply:
If the reason is that the data type for all format_data_ Tensors MUST be the same (so they're implicitly aligned as long as the first one is), it would be good to state that in a comment.

Reply:
I just want to process 32 and 64 bit data aligned on 64 boundary  and not on the arbitrary byte right after data.